### PR TITLE
nixos/docker-registry: add configuration options to support image deletions

### PR DESCRIPTION
**Pull request by Ma27**
###### Motivation for this change

The main purpose of this PR originally prepared by @ironpinguin and finished by myself was to allow deletions of image revisions in a docker registry.
The original module used the sample configuration provided by `docker/distribution` which disabled this feature. We decided to generate a JSON-based config based on a Nix object that can be filled with values from the configuration. Furthermore the testcase ensures now that images can be deleted.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/modules/services/misc/docker-registry.nix [hunk 1]
```diff
@@ -41,15 +102,11 @@ in {
...
-
-      environment = {
-        REGISTRY_HTTP_ADDR = "${cfg.listenAddress}:${toString cfg.port}";
-        REGISTRY_STORAGE_FILESYSTEM_ROOTDIRECTORY = cfg.storagePath;
-      } // cfg.extraConfig;
-
-      script = ''
-        ${pkgs.docker-distribution}/bin/registry serve \
-          ${pkgs.docker-distribution.out}/share/go/src/github.com/docker/distribution/cmd/registry/config-example.yml
+      environment = cfg.extraConfig;
```
Comment (Ma27) [inline nixos/modules/services/misc/docker-registry.nix]: actually we should drop this and merge the `extraConfig` with the configuration tree

File: nixos/modules/services/misc/docker-registry.nix [hunk 2]
```diff
@@ -27,6 +64,30 @@ in {
...
+
+    redisUrl = mkOption {
+      type = types.str;
+      default = "localhost:6479";
+      description = "Set redis host and port.";
+    };
+
+    redisPassword = mkOption {
+      type = types.str;
+      default = "asecret";
```
Comment (Ma27) [inline nixos/modules/services/misc/docker-registry.nix]: we should not use set a secret as default here

**Review by Ma27 — COMMENTED** [review]
just found two issues, will fix them and document the breaking change with `extraConfig` in the release notes

File: nixos/tests/docker-registry.nix [hunk 3]
```diff
@@ -43,7 +44,20 @@ import ./make-test.nix ({ pkgs, ...} : {
...
+
+    $registry->systemctl("start docker-registry-garbage-collect.service");
+    $registry->waitUntilFails("systemctl status docker-registry-garbage-collect.service");
+    $registry->waitForUnit("docker-registry.service");
+
+    $registry->fail(
+      'ls -l /var/lib/docker-registry/docker/registry/v2/blobs/sha256/*/*/data'
+    );
+
+    $client1->succeed("docker push registry:8080/scratch");
```
Comment (nlewo) [inline nixos/tests/docker-registry.nix]: Why are you pushing the image again?
Wouldn't be sufficient to do this test just before starting the garbage collector (around line 40)?

Comment (Ma27) [reply nixos/tests/docker-registry.nix]: we wanted to ensure that after running gc and confirming that all blobs were purged that the registry is still functional akd can receive new images

Comment (nlewo) [reply nixos/tests/docker-registry.nix]: ok, thanks

**Comment by Ma27** [comment]
@globin @fpletz I rebased the patches onto the latest master (to resolve some merge conflicts in the release notes), could you have a look at this pls? :)

File: nixos/modules/services/misc/docker-registry.nix [hunk 4]
```diff
@@ -27,29 +66,55 @@ in {
...
+      type = types.bool;
+      default = false;
+      description = "Enable delete for manifests and blobs.";
+    };
+
+    enableRedisCache = mkEnableOption "redis as blob cache";
+
+    redisUrl = mkOption {
+      type = types.str;
+      default = "localhost:6479";
```
Comment (globin) [inline nixos/modules/services/misc/docker-registry.nix]: 6379?

**Review by globin — CHANGES_REQUESTED** [review]

**Comment by Ma27** [comment]
@globin good catch - thanks!

**Review by globin — APPROVED** [review]

**Pull request merged**